ENH: Allow arbitrary dtypes in binomial

bashtage opened the pull request:
Allow arbitrary dtypes as long as the values are in range
for long

closes #2011

bashtage committed:
ENH: Allow arbitrary dtypes in binomial

Allow arbitrary dtypes as long as the values are in range
for long

closes #2011

eric-wieser commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -3698,7 +3698,8 @@ cdef class RandomState:
             Parameter of the distribution, >= 0. Floats are also accepted,
-            but they will be truncated to integers.
+            but they will be truncated to integers. All values must be less
+            than or equal to np.iinfo(np.long).max
```
`np.long` is not the same as NPY_LONG. the former is simply the built-in `long`, or `int` in python 3

bashtage commented:
I figured this out in the code part, but what is the right way to
communicate the bound to the end user? Iinfo(int).max?

On Tue, Aug 22, 2017, 12:43 Eric Wieser <notifications@github.com> wrote:

> *@eric-wieser* commented on this pull request.
> ------------------------------
>
> In numpy/random/mtrand/mtrand.pyx
> <https://github.com/numpy/numpy/pull/9589#discussion_r134455351>:
>
> > @@ -3698,7 +3698,8 @@ cdef class RandomState:
>          ----------
>          n : int or array_like of ints
>              Parameter of the distribution, >= 0. Floats are also accepted,
> -            but they will be truncated to integers.
> +            but they will be truncated to integers. All values must be less
> +            than or equal to np.iinfo(np.long).max
>
> np.long is not the same as NPY_LONG. the former is simply the built-in
> long, or int in python 3
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/9589#pullrequestreview-57747378>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AFU5RSv9zy-m4E2Y3MVUvx0WGZblBVeiks5sar7NgaJpZM4O-fj8>
> .
>

eric-wieser commented:
Possibly `iinfo('l')`? Worth checking if that's correct.

Also, I suspect the tests will fall on win32, where long is only an int32 (or at least, `builtins.int` is)

bashtage commented:
Using 'np.iinfo('l').max` seems to work across all all platforms.  It is also used for another constant for the maximum Poisson intensity, which indicates it is likely correct.

eric-wieser commented:
I'd argue that this fix is not desirable, as it changes a dtype-based error (easy to catch with tests) with a value-based one.

bashtage commented:
The function is a bit uneven in its treatment of inputs.

```python
np.random.binomial([3.0,10.2],0.5)
Out[26]: array([2, 7])

np.random.binomial(np.array([3.0,10.2]),0.2)
TypeError: Cannot cast array data from dtype('float64') to dtype('int32') according to the rule 'safe'

np.random.binomial(np.array([3.0,10.2],dtype=np.int32),0.2)
Out[28]: array([1, 6])

np.random.binomial(np.array([3.0,10.2],dtype=np.int64),0.2)
Cannot cast array data from dtype('int64') to dtype('int32') according to the rule 'safe'
```

bashtage committed:
REF: Switch to OverflowError

Switch to OverflowError to inform cause of issue since this is not
technically a ValueError as n has no natural upper bound

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -3771,16 +3772,21 @@ cdef class RandomState:
         cdef double fp
+        
+        on = <ndarray>np.array(n)
+        if np.any(np.greater(n, np.iinfo('l').max)):
```
So this converts `n` to a numpy array, which is what `on` already is.

charris commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -3771,16 +3772,21 @@ cdef class RandomState:
+        if np.any(np.greater(n, np.iinfo('l').max)):
+            raise OverflowError('n is out of range for C long '
+                                '(max : {0})'.format(np.iinfo('l').max))
+        if np.any(np.less(n, 0)):
```
See above.

charris commented:
I'm in general agreement with @eric-wieser here, and would actually prefer that the caller deal with the values before passing them. @rkern has also suggested making a pass through all of np.random at sometime in the future and dealing systematically with the long parameters.

bashtage commented:
This is probably the wrong way -- converting to `int64` -- especially the ability to accept `int64` -- is a much more sensible approach IMO.
